SI-8631 Treat `A with Sealed` as enumerable for pattern matching

retronym opened the pull request:
Enumerate the subtypes of sealed parent types of refinement
types, and filter those that conform to the refinement type.

Such types can crop up easily when LUB-bing case classes which
add `Product with Serializable` to the mix.

retronym committed:
SI-8631 Treat `A with Sealed` as enumerable for pattern matching

Enumerate the subtypes of sealed parent types of refinement
types, and filter those that conform to the refinement type.

Such types can crop up easily when LUB-bing case classes which
add `Product with Serializable` to the mix.

retronym commented:
As suggested by @aloiscochard: https://twitter.com/aloiscochard/status/471586327163449344

Review by @adriaanm

aloiscochard commented:
Actually kudos to @ijuma for making me realize that it should be enumerable :-)

Awesome @retronym, thanks a lot!

ijuma commented:
Thanks indeed. :)

adriaanm commented on src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala:
```diff
@@ -105,6 +105,9 @@ trait TreeAndTypeAnalysis extends Debugging {
         case modSym: ModuleClassSymbol =>
           Some(List(tp))
         // make sure it's not a primitive, else (5: Byte) match { case 5 => ... } sees no Byte
+        case sym: RefinementClassSymbol
```
I think it would be nice to see about combining `!sym.isSealed` (below) && `!sym.info.parents.exists(_.typeSymbol.isSealed)` and treating the alternative uniformly in `case sym =>`

retronym replied:
Yep, I agree that would be an improvement. I'll do that this week.

adriaanm commented on src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala:
```diff
@@ -105,6 +105,9 @@ trait TreeAndTypeAnalysis extends Debugging {
         // make sure it's not a primitive, else (5: Byte) match { case 5 => ... } sees no Byte
+        case sym: RefinementClassSymbol
+          if sym.info.parents.exists(_.typeSymbol.isSealed) =>
+          Some(tp.parents.flatMap(parent => enumerateSubtypes(parent).getOrElse(Nil).filter(_ <:< tp)))
```
why do we need the `filter(_ <:< tp)`?

retronym replied:
`tp` might have decls that are not satisfied by the descendants of the parents. I have a few test variations that show that in action.

adriaanm replied:
Ah yes, since we're only enumerating parts of the intersection. Would be good to add that as a comment.

adriaanm commented:
@retronym, the idea LGTM, but I'd like to see some consolidation of the sealedness logic

retronym commented:
Re-review, please, by @adriaanm

retronym committed:
Refactor `enumerateSubtypes`

  - reorder to make "unenumerable" the last case
  - reunite an orphaned comment with its loved ones.

adriaanm commented:
LGTM -- thanks for addressing my nitpicks!

retronym committed:
Remove now-redundant condition in sealedness checking

It arrived in c1f8dbc, but is no longer needed after user-defined
value classes were introduced in 2.10.0.
